File: src/experiments/compare_spmf.py
import sys
import os
from pathlib import Path

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))
from utils import read_spmf_file

def parse_spmf_output(filepath):
    """Đọc file output của SPMF hoặc GenMax"""
    results = {}
    with open(filepath, 'r', encoding='utf-8') as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith('#'):
                continue
            if '#SUP:' in line:
                parts = line.split('#SUP:')
                itemset_str = parts[0].strip()
                sup = int(parts[1].strip())
                itemset = tuple(sorted(map(int, itemset_str.split())))
                results[itemset] = sup
    return results

def compare_results(our_file, spmf_file):
    our = parse_spmf_output(our_file)
    spmf = parse_spmf_output(spmf_file)
    
    print(f"\n=== SO SÁNH CORRECTNESS ===")
    print(f"Our GenMax   : {len(our)} MFI")
    print(f"SPMF         : {len(spmf)} MFI")
    
    common = set(our.keys()) & set(spmf.keys())
    only_our = set(our.keys()) - set(spmf.keys())
    only_spmf = set(spmf.keys()) - set(our.keys())
    
    print(f"Khớp nhau    : {len(common)}")
    print(f"Chỉ GenMax   : {len(only_our)}")
    print(f"Chỉ SPMF     : {len(only_spmf)}")
    
    if len(our) == len(spmf) and len(only_our) == 0 and len(only_spmf) == 0:
        print("✅ HOÀN HẢO - Kết quả khớp 100% với SPMF!")
    else:
        print("⚠️  Có sai lệch, cần kiểm tra code.")

if __name__ == "__main__":
    # Ví dụ sử dụng:
    compare_results("results/chess_mfi_our.txt", "results/chess_mfi_spmf.txt")

File: src/experiments/compare_spmf.jl
function parse_spmf_output(filepath::String)
    results = Dict{Vector{String}, Int}()
    for line in eachline(filepath)
        line = strip(line)
        if isempty(line) || startswith(line, "#") && !occursin("#SUP:", line)
            continue
        end
        if occursin("#SUP:", line)
            parts = split(line, "#SUP:")
            itemset_str = strip(parts[1])
            sup = parse(Int, strip(parts[2]))
            
            # Lưu ý mảng item có thể là dạng số hoặc chữ, lưu ở dạng chuỗi thống nhất
            itemset = sort(string.(split(itemset_str)))
            results[itemset] = sup
        end
    end
    return results
end

function compare_results(our_file::String, spmf_file::String)
    our = parse_spmf_output(our_file)
    spmf = parse_spmf_output(spmf_file)
    
    println("\n=== SO SÁNH CORRECTNESS ===")
    println("Kết quả do GenMax custom sinh : $(length(our)) MFI")
    println("Kết quả do tool SPMF java sinh: $(length(spmf)) MFI")
    
    our_keys = Set(keys(our))
    spmf_keys = Set(keys(spmf))
    
    common = intersect(our_keys, spmf_keys)
    only_our = setdiff(our_keys, spmf_keys)
    only_spmf = setdiff(spmf_keys, our_keys)
    
    println("Khớp nhau hoàn toàn : $(length(common))")
    println("Chỉ có bên GenMax   : $(length(only_our))")
    println("Chỉ có bên SPMF     : $(length(only_spmf))")
    
    if length(our) == length(spmf) && isempty(only_our) && isempty(only_spmf)
        println("HOÀN HẢO - Kết quả code của bạn khớp 100% với SPMF!")
    else
        println("Cảnh báo: Có sự sai lệch, kết quả chưa chuẩn.")
    end
end

if abspath(PROGRAM_FILE) == @__FILE__
    if length(ARGS) >= 2
        compare_results(ARGS[1], ARGS[2])
    else
        println("Sử dụng: julia compare_spmf.jl <our_output.txt> <spmf_output.txt>")
    end
end
